File: app/src/main/java/com/candy/capture/core/BackStackManager.java
package com.candy.capture.core;

import android.app.Activity;

import java.util.Iterator;
import java.util.Stack;

/**
 * Created by zhanghq on 2016/11/17.
 */

public class BackStackManager {

    private static BackStackManager backStackManager;
    private Stack<Activity> mActivityStack;

    private BackStackManager() {
        mActivityStack = new Stack<>();
    }

    public static BackStackManager getInstance() {
        if (backStackManager == null) {
            backStackManager = new BackStackManager();
        }
        return backStackManager;
    }

    public void push(Activity activity) {
        mActivityStack.push(activity);
    }

    public void removeSpecificActivity(Activity activity) {
        mActivityStack.remove(activity);
    }

    public void removeSpecificActivityByClass(Class<?> cls) {
        Iterator<Activity> iterator = mActivityStack.iterator();
        Activity activity = null;
        while (iterator.hasNext()) {
            activity = iterator.next();
            if (activity.getClass().equals(cls)) {
                activity.finish();
                iterator.remove();
            }
        }
    }
}

File: app/src/main/java/com/candy/capture/core/BackStackManager.kt
package com.candy.capture.core

import android.app.Activity
import java.util.*

/**
 * @Description
 * @Author zhanghaiqiang
 * @Date 2020/9/30 11:18
 */
class BackStackManager private constructor() {

    private var mActivityStack: Stack<Activity> = Stack()

    companion object {
        private var backStackManager: BackStackManager? = null

        fun getInstance(): BackStackManager {
            if (backStackManager == null) {
                backStackManager = BackStackManager()
            }
            return backStackManager!!
        }
    }

    fun push(activity: Activity) {
        mActivityStack.push(activity)
    }

    fun removeSpecificActivity(activity: Activity) {
        mActivityStack.remove(activity)
    }

    fun removeSpecificActivityByClass(cls: Class<*>) {
        val iterator = mActivityStack.iterator()
        var activity: Activity? = null
        while (iterator.hasNext()) {
            activity = iterator.next()
            if (activity.javaClass == cls) {
                activity.finish()
                iterator.remove()
            }
        }
    }
}